quipucords/scanner/normalizer.py

```diff
@@ -0,0 +1,225 @@
+"""normalizer module."""
+
+from dataclasses import dataclass
+from logging import getLogger
+from typing import Any
+
+logger = getLogger(__name__)
+
+
+class NormalizerMeta(type):
+    """
+    Type for Normalizer classes.
+
+    Customize class creation to provide the necessary magic for Normalizers.
+
+    This metaclass is meant to be used ONLY on BaseNormalizer.
+    """
+
+    def __new__(cls, name, bases, attrs):
+        """Blueprint for Normalizer classes."""
+        # we need fields to be accessible as a class attribute. This is the WHOLE reason
+        # for the existence of this metaclass (I wish python had a simpler api for this
+        # type of thing).
+        # Usually just creating a attribute on a base class would be enough
+        # for this behavior, but not when you are dealing with dicts or lists.
+        # (Creating 'fields' as a normal attribute on 'BaseNormalizer' class would
+        # result on it being shared between the base class and its subclasses.)
+        attrs["fields"] = {}
+        return super().__new__(cls, name, bases, attrs)
+
+
+class BaseNormalizer(metaclass=NormalizerMeta):
+    """Base class for Normalizer classes."""
+
+    # must be set for other implementations
+    source_type = None
+
+    def __init__(self, raw_facts: dict, server_id: str = None) -> None:
+        """Create a normalizer instance.
+
+        :param raw_facts: dict of raw facts from inspection phase
+        :param server_id: server_id of quipucords server that collected these
+        raw facts.
+        """
+        self.raw_facts = raw_facts
+        self._metadata = {}
+        self.server_id = server_id
+
+    @property
+    def facts(self):
+        """Return normalized facts."""
+        assert hasattr(
+            self, "_normalized"
+        ), "Call 'normalize()' before asking for facts."
```

Comment:
I forgot about that. Some linters even complain about asserts in the middle of non test code because of that...

Reply:
this is where we need to iterate over fields following the order they were created

https://github.com/quipucords/quipucords/blob/fd552f9f95af83c617f7d152d823cc9528f860fe/quipucords/scanner/normalizer.py#L73

And this is required because I considered preserving order + checking at creation time if the dependency field do exist (implementation https://github.com/quipucords/quipucords/blob/fd552f9f95af83c617f7d152d823cc9528f860fe/quipucords/scanner/normalizer.py#L201-L205 + test https://github.com/quipucords/quipucords/blob/911326a4c7cd4446a6b4c5dc48b06e9cad7b868b/quipucords/tests/scanner/test_normalizer.py#L28-L38) the best heuristics to ensure field dependencies would be respected.

I like this idea because if a normalizer is "incorrectly" implemented it would fail ASAP - and is relatively simple.

But if you have an idea that would completely avoid the metaclass I'm all ears =D

Reply:
exluding tests from selection, I'll have to change some code to remove assertions from production code
```
ruff check --select=S101 .
quipucords/api/source/serializer.py:516:17: S101 Use of `assert` detected
quipucords/scanner/manager.py:43:9: S101 Use of `assert` detected
quipucords/scanner/manager.py:143:9: S101 Use of `assert` detected
quipucords/scanner/openshift/api.py:214:9: S101 Use of `assert` detected
quipucords/scanner/openshift/entities.py:59:9: S101 Use of `assert` detected
quipucords/scanner/openshift/entities.py:60:9: S101 Use of `assert` detected
Found 6 errors.
```

Reply:
While I haven't checked, I assumed this is not the only place where `assert` is used. So that was more of an off-hand comment "hey, maybe we shouldn't be doing that", instead of hard requirement to change your PR.

Reply:
done =D